rust/kernel/lib.rs

```diff
@@ -69,6 +70,34 @@ pub trait Module: Sized + Sync {
     fn init(module: &'static ThisModule) -> error::Result<Self>;
 }
 
+/// A module that is pinned and initialised in-place.
+pub trait InPlaceModule: Sync + Send {
+    /// The type that implements the pinned initialisation of the module.
+    type Init: init::PinInit<Self, error::Error>;
+
+    /// Creates an initialiser for the module.
+    ///
+    /// It is called when the module is loaded.
+    fn init(module: &'static ThisModule) -> error::Result<Self::Init>;
```

Comment:
Why are you using `Result<impl Init<Self, Error>>` here? There are then two places that can fail:
- the creation of the initializer
- the initialization itself.

Just wanted to check that you actually want that. Not really sure if we want that, I would need to view some examples to  decide for myself.